app/src/main/java/io/zenandroid/onlinego/data/repositories/PuzzleRepository.kt

```diff
@@ -0,0 +1,164 @@
+
+        return dao.getPuzzle(id)
+                .doOnNext { Log.d("PuzzleRepository", it.toString()) }
+                .distinctUntilChanged()
+    }
+
+    fun getPuzzleRating(id: Long): Flowable<PuzzleRating> {
+        disposable += restService.getPuzzleRating(id)
+                .map { it.copy(puzzleId = id) }
+                .subscribe(this::savePuzzleRatingToDB, this::onError)
+
+        return dao.getPuzzleRating(id)
+                .onErrorReturnItem(PuzzleRating(
+                    puzzleId = id,
+                    rating = -1,
+                ))
+                .doOnNext { Log.d("PuzzleRepository", it.toString()) }
+                .distinctUntilChanged()
+    }
+
+    fun ratePuzzle(id: Long, rating: Int): Single<PuzzleRating> =
+        restService.ratePuzzle(id, rating)
+                .doOnComplete { Log.d("PuzzleRepository", "rate $id: $rating") }
+                .andThen(Single.just(PuzzleRating(puzzleId = id, rating = rating)))
+                .also {
+                    it.subscribe(this::savePuzzleRatingToDB, this::onError)
+                        .addToDisposable(disposable)
+                }
+
+    fun getPuzzleSolution(id: Long): Flowable<List<PuzzleSolution>> {
+        disposable += restService.getPuzzleSolutions(id)
+                .subscribe(this::savePuzzleSolutionsToDB, this::onError)
+
+        return dao.getPuzzleSolution(id)
+                .onErrorReturnItem(emptyList())
+                .doOnNext { Log.d("PuzzleRepository", it.toString()) }
+                .distinctUntilChanged()
+    }
+
+    fun getPuzzleCollectionSolutions(): Flowable<Map<Long, Int>> {
+        return dao.getPuzzleCollectionSolutions()
+            .onErrorReturnItem(emptyList())
+            .doOnNext { Log.d("PuzzleRepository", it.toString()) }
+            .map { it.sortedBy { it.collectionId } }
+            .distinctUntilChanged()
+            .map { it.associateBy({ it.collectionId }, { it.count }) }
+    }
+
+    fun markPuzzleSolved(id: Long, record: PuzzleSolution): Single<PuzzleSolution> =
+        restService.markPuzzleSolved(id, record)
+                .doOnComplete { Log.d("PuzzleRepository", "solve $id: $record") }
+                .andThen(Single.just(record))
+                .also {
+                    it.map(::listOf).subscribe(this::savePuzzleSolutionsToDB, this::onError)
+                        .addToDisposable(disposable)
+                }
+
+    private fun saveCollectionsToDB(list: List<PuzzleCollection>) {
+        dao.insertPuzzleCollections(list)
+    }
+
+    private fun savePuzzlesToDB(list: List<Puzzle>) {
+        dao.insertPuzzles(list)
+    }
+
+    private fun savePuzzleRatingToDB(item: PuzzleRating) {
+        dao.insertPuzzleRating(item)
+    }
+
+    private fun savePuzzleSolutionsToDB(list: List<PuzzleSolution>) {
+        dao.insertPuzzleSolutions(list)
+    }
+
+    private fun onError(error: Throwable) {
+        Log.e("PuzzleRepository", error.message, error)
```

Comment:
please use `recordException(error)` so that the error is captured by Crashlytics.